bundles/org.openhab.binding.enigma2/src/main/java/org/openhab/binding/enigma2/internal/Enigma2Client.java

```diff
@@ -0,0 +1,350 @@
+    private final String host;
+    private boolean power;
+    private String channel = "";
+    private String title = "";
+    private String description = "";
+    private String answer = "";
+    private int volume = 0;
+    private boolean mute;
+    private boolean online;
+    private boolean initialized = false;
+    private boolean asking = false;
+    private LocalDateTime lastAnswerTime = LocalDateTime.of(2020, 1, 1, 0, 0); // Date in the past
+    private Enigma2HttpClient enigma2HttpClient;
+
+    public Enigma2Client(String host, @Nullable String user, @Nullable String password, int requestTimeout) {
+        this.enigma2HttpClient = new Enigma2HttpClient(requestTimeout);
+        if (StringUtils.isNotEmpty(user) && StringUtils.isNotEmpty(password)) {
+            this.host = "http://" + user + ":" + password + "@" + host;
+        } else {
+            this.host = "http://" + host;
+        }
+    }
+
+    public boolean refresh() {
+        boolean wasOnline = online;
+        refreshPower();
+        if (!wasOnline && online) {
+            // Only refresh all services if the box changed from offline to online and power is on
+            // because it is a performance intensive action.
+            refreshAllServices();
+        }
+        refreshChannel();
+        refreshEpg();
+        refreshVolume();
+        refreshAnswer();
+        return online;
+    }
+
+    public void refreshPower() {
+        Optional<Document> document = transmitWithResult(PATH_POWER);
+        if (document.isPresent()) {
+            online = true;
+            processPowerResult(document.get());
+        } else {
+            online = false;
+            power = false;
+        }
+        initialized = true;
+    }
+
+    public void refreshAllServices() {
+        if (power || channels.isEmpty()) {
+            transmitWithResult(PATH_ALL_SERVICES).ifPresent(this::processAllServicesResult);
+        }
+    }
+
+    public void refreshChannel() {
+        if (power) {
+            transmitWithResult(PATH_CHANNEL).ifPresent(this::processChannelResult);
+        }
+    }
+
+    public void refreshAnswer() {
+        if (asking) {
+            transmitWithResult(PATH_ANSWER).ifPresent(this::processAnswerResult);
+        }
+    }
+
+    public void refreshVolume() {
+        if (power) {
+            transmitWithResult(PATH_VOLUME).ifPresent(this::processVolumeResult);
+        }
+    }
+
+    public void refreshEpg() {
+        if (power) {
+            Optional.ofNullable(channels.get(channel))
+                    .flatMap(name -> transmitWithResult(PATH_EPG + UrlEncoded.encodeString(name)))
+                    .ifPresent(this::processEpgResult);
+        }
+    }
+
+    private Optional<Document> transmitWithResult(String path) {
+        try {
+            Optional<String> xml = transmit(path);
+            if(xml.isPresent()) {
+                DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
```

Comment:
I do believe the point of this factory is that you do not need to create it each time, I would suggest to either construct it in the constructor or lazy on calling some method but please do not recreate the factory each time.

Reply:
Fixed